// generate-license-file/dist/models/license.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.License = void 0;
const BULLET = " - ";
const PREFIX = "The following npm package may be included in this product:";
const PREFIX_PLURAL = "The following npm packages may be included in this product:";
const MIDFIX = "This package contains the following license and notice below:";
const MIDFIX_PLURAL = "These packages each contain the following license and notice below:";
class License {
    constructor(content, dependencies) {
        this.content = content;
        this.dependencies = dependencies;
    }
    format(EOL) {
        const formattedText = this.prefix(EOL) +
            this.formatDependencies(EOL) +
            this.midfix(EOL) +
            this.normalizeLineEndings(this.content.trim(), EOL);
        return formattedText;
    }
    prefix(EOL) {
        if (this.dependencies.length === 1) {
            return PREFIX + EOL + EOL;
        }
        return PREFIX_PLURAL + EOL + EOL;
    }
    formatDependencies(EOL) {
        const formattedText = this.dependencies.map(dependency => {
            return BULLET + dependency + EOL;
        });
        return formattedText.join("");
    }
    midfix(EOL) {
        if (this.dependencies.length === 1) {
            return EOL + MIDFIX + EOL + EOL;
        }
        return EOL + MIDFIX_PLURAL + EOL + EOL;
    }
    normalizeLineEndings(text, EOL) {
        return text.replace(License.lineEndingRegex, EOL);
    }
}
exports.License = License;
License.lineEndingRegex = /\r\n|\r|\n/g;
